### PR TITLE
Longest Substring Length

### LINKED ISSUE #51: Longest Substring Without Repeating Characters
### Description:
Given a string `s`, find the length of the longest substring without repeating characters.

### Input:
- A string `s` containing English letters, digits, symbols, and spaces.

### Output:
- An integer representing the length of the longest substring of `s` that contains no repeating characters.

#### Example 1:
Input: `s = "abcabcbb"`  
Output: `3`  

#### Example 2:
Input: `s = "bbbbb"`  
Output: `1`  

#### Example 3:
Input: `s = ""`  
Output: `0`  

---

[Leetcode](https://leetcode.com/problems/longest-substring-without-repeating-characters/solutions/)

### PR DESCRIPTION
**Pull request by jola-ds**
Fixes #51 
---
name: solution review 
about: A template PR for code review with a checklist
---

# Behavior
This function takes in a string s and returns an integer representing the length of the longest substring of s that contains no repeating characters.

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/longest_substring_length.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+A module to find the length of the longest substring in a given string without repeating characters.
```
Comment (mojtabafayyaz) [inline solutions/longest_substring_length.py:5]: Could you add a description of the strategy?

Comment (jola-ds) [reply solutions/longest_substring_length.py:5]: In the module docstring?

**Review by mojtabafayyaz — CHANGES_REQUESTED** [review]
**Dear Jola,**

Great job! The logic is solid, and the test cases cover various scenarios, including edge cases and error handling. The code is clear and well-organized. However, there are a few areas for improvement to make it even better.

 **Suggested Improvements:**

  - Update the docstring to replace `AssertionError` with `TypeError` for type validation, as it's more appropriate for handling type-related issues (e.g., when the input is not a string).
  
  - In the code, replace `AssertionError` with `TypeError` when performing type validation for the string input.

  - Add a description of the sliding window strategy in the function docstring to provide additional context for future readers and explain the approach used in the code.

You're doing an excellent job with the code, and with these improvements, your implementation will be even clearer and more aligned with best practices.

File: solutions/longest_substring_length.py (hunk 1 above)
Comment (mojtabafayyaz) [reply solutions/longest_substring_length.py:5]: Yes.

**Comment by jola-ds** [comment]
Thanks for the feedback @mojtabafayyaz 

I would prefer to keep the AssertionError as it is in the code, because the code review checklist specifically mentions that the code should include defensive assertions. Additionally, the AssertionError effectively validates the input and ensures the expected conditions are met.

For the second suggestion, I noticed you marked the checklist about not including 'strategy' in the documentation. However, your review suggests adding the strategy to the function docstring and I'm a bit confused. My understanding of the checklist is that strategy should not be included in documentation, could you please clarify? I want to ensure I'm following the expectations correctly.

**Review by mojtabafayyaz — APPROVED** [review]
Dear Jola,

Apologies for the misunderstanding regarding the brief description of the strategy in the module docstring. You are absolutely right — the checklist strictly discourages describing the strategy in the documentation.  Thanks for clarification..! 

Regarding the AssertionError, there's nothing wrong with using it as you have. My suggestion was simply that TypeError might be more specific for type validation, and could be a more precise way of handling type-related issues. However, if you prefer to keep the AssertionError, I completely understand, as it still serves as a valid defensive check. My suggestions were only meant to offer some options to enhance clarity, which is why I marked them as optional in the checklist.

Thanks for your work on this, and feel free to proceed as you see fit!

**Comment by jola-ds** [comment]
Thanks, Fayyaz!
